koap/src/main/kotlin/ByteArrayReader.kt

```diff
@@ -0,0 +1,84 @@
+package com.juul.koap
+
+internal inline fun <T> ByteArray.withReader(
+    startIndex: Int = 0,
+    endIndex: Int = size,
+    action: ByteArrayReader.() -> T
+): T = action.invoke(ByteArrayReader(this, startIndex, endIndex))
+
+internal fun ByteArray.reader(
+    startIndex: Int = 0,
+    endIndex: Int = size
+): ByteArrayReader = ByteArrayReader(this, startIndex, endIndex)
+
+/**
+ * All numbers are read in network byte-order (Big endian).
+ *
+ * @param startIndex the start of the range (inclusive), must be in `0..bytes.size`
+ * @param endIndex the end of the range (exclusive), must be in `startIndex..bytes.size`
+ */
+internal class ByteArrayReader(
+    private val bytes: ByteArray,
+    startIndex: Int = 0,
+    private val endIndex: Int = bytes.size
+) {
+
+    var index = startIndex
+    fun exhausted(): Boolean = index >= endIndex
+
+    private fun checkIndex() {
+        check(index < endIndex) { "Cannot read at index $index, >= endIndex of $endIndex" }
+    }
+
+    /** Reads 1-byte to acquire an 8-bit unsigned int. */
+    fun readUByte(): Int {
+        checkIndex()
+        return bytes[index++].toInt() and 0xFF
+    }
+
+    /** Reads 2-bytes in [ByteArray] receiver to acquire a 16-bit unsigned int. */
+    fun readUShort(): Int {
+        checkIndex()
+        return ((bytes[index++].toInt() and 0xFF) shl 8) or readUByte()
+    }
+
+    /** Reads 3-bytes in [ByteArray] receiver to acquire a 24-bit unsigned int. */
+    fun readUInt24(): Int {
+        checkIndex()
+        return ((bytes[index++].toInt() and 0xFF) shl 16) or readUShort()
+    }
+
+    /** Reads 4-bytes in [ByteArray] receiver to acquire a 32-bit unsigned int. */
+    fun readUInt(): Long {
+        checkIndex()
+        return ((bytes[index++].toLong() and 0xFF) shl 24) or readUInt24().toLong()
+    }
+
+    /** Reads 8-bytes in [ByteArray] receiver to acquire a 64-bit signed int. */
+    fun readLong(): Long {
```

Comment:
No change request with this comment, just want to get your thoughts on the following:

Do you think it's worth checking for the specific number of bytes you expect to read? Or is that a user problem?

In my opinion, I think it would be a nice-to-have if this class threw its own exception stating that there aren't enough bytes to read for the requested read operation. I don't think it would provide a huge gain for the user over seeing an `IndexOutOfBoundsException`, but I think it would be good to have our own sanity check to throw a specific exception per read operation if there aren't enough bytes to read.

Reply:
Definitely agree. We should provide informative exceptions. That's what `checkIndex` does, but the exception message is a bit cryptic, so I'll fix that.

Reply:
Updated `checkIndex` to hopefully be more readable?

```kotlin
private fun checkIndex() {
    if (index >= endIndex) {
        throw IndexOutOfBoundsException(
            "Cannot read when index is at or beyond endIndex (index=$index, endIndex=$endIndex)"
        )
    }
}
```